src/main/scala/zio/ftp/UnsecureFtp.scala

```diff
@@ -35,37 +35,34 @@ final private class UnsecureFtp(unsafeClient: Client) extends FtpAccessors[Clien
     execute(c => Option(c.mlistFile(path))).map(_.map(FtpResource.fromFtpFile(_)))
 
   def readFile(path: String, chunkSize: Int = 2048): ZStream[Blocking, IOException, Byte] = {
-    val terminate = execute(_.completePendingCommand())
-      .flatMap {
-        case false =>
-          ZIO.fail(new IOException(s"Cannot finalize the file transfer and complete to read the entire file $path."))
-        case _     => ZIO.unit
+    val terminate = execute(_.completePendingCommand()).flatMap {
+      case false =>
+        ZIO.fail(
+          FileTransferIncompleteError(s"Cannot finalize the file transfer and completely read the entire file $path.")
+        )
+      case _     => ZIO.unit
+    }
+
+    val inputStream =
+      execute(c => Option(c.retrieveFileStream(path)))
+        .flatMap(
+          _.fold[ZIO[Any, InvalidPathError, InputStream]](
+            ZIO.fail(InvalidPathError(s"File does not exist $path"))
+          )(
+            ZIO.succeed(_)
+          )
+        )
+
+    ZStream
+      .fromInputStreamManaged(
+        ZManaged
+          .make(inputStream)(a => UIO(a.close()))
```

Comment:
you can replace by 

```scala
ZStream.managed(ZManaged.fromAutoCloseable(Task(is))).mapError(e => new IOException(e.getMessage, e))
```

Reply:
turns out there's an even easier one - `ZStream.fromInputStream[Effect]` that wraps it in a managed and closes the stream